### PR TITLE
change cs.StartTime to cs.CPU.Time in addPodCPUMemoryStats

### PR DESCRIPTION
**Pull request by Zheaoli**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes # https://github.com/kubernetes/kubernetes/issues/93988

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by dashpole** [comment]
/ok-to-test

I think you will have to update the unit test for this

/approve

**Comment by Zheaoli** [comment]
@dashpole thanks to remind me for that. I'm working on it

**Comment by Zheaoli** [comment]
It seems all test pass, don't need extra unit tests. PTAL  @dashpole

**Comment by dashpole** [comment]
I think that means it actually just isn't unit tested at all...

Looking at https://github.com/kubernetes/kubernetes/blob/fd74333a971e2048b5fb2b692a9e043483d63fba/pkg/kubelet/stats/stats_provider_test.go#L503, we don't even set timestamps on the CPU metrics returned by the fake.  Can you update the tests so that we are actually testing this?

**Comment by Zheaoli** [comment]
@dashpole Sure, I'll update the test ASAP

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 1]
```diff
@@ -464,7 +464,7 @@ func (p *criStatsProvider) addPodCPUMemoryStats(
 			ps.CPU = &statsapi.CPUStats{}
 		}
 
-		ps.CPU.Time = cs.StartTime
+		ps.CPU.Time = cs.CPU.Time
```
Comment (krmayankk) [inline pkg/kubelet/stats/cri_stats_provider.go:467]: @Zheaoli  does this require a doc bug fix ? Where do these stats show up  for users ?

Comment (Zheaoli) [reply pkg/kubelet/stats/cri_stats_provider.go:467]: I'm not sure about the doc. I will check it again

**Comment by dims** [comment]
Need to remove WIP from title

/retest

**Comment by Zheaoli** [comment]
@dims I work on fix the unit test. but maybe I need some help

**Comment by SergeyKanzhelev** [comment]
@Zheaoli do you still plan to update unit tests to validate this field?

**Comment by Zheaoli** [comment]
@SergeyKanzhelev If it's OK I will use another PR to update the related unit test.

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 1 above)
Comment (dashpole) [reply pkg/kubelet/stats/cri_stats_provider.go:467]: I don't think we document this that specifically.

Comment (dashpole) [reply pkg/kubelet/stats/cri_stats_provider.go:467]: This is for creating the Summary API.

**Comment by dashpole** [comment]
@Zheaoli we will likely want to cherrypick this to previous releases, and it would be easier to do so with the unit test included.  If you can push what you have that is failing, we may be able to help debug failures.

File: pkg/kubelet/stats/cri_stats_provider_test.go [hunk 2]
```diff
@@ -421,87 +421,89 @@ func TestCRIListPodCPUAndMemoryStats(t *testing.T) {
 	)
 
 	stats, err := provider.ListPodCPUAndMemoryStats()
-	assert := assert.New(t)
-	assert.NoError(err)
-	assert.Equal(4, len(stats))
+	assertObject := assert.New(t)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/stats/cri_stats_provider_test.go:424]: curious, why to rename the variable?

Comment (SergeyKanzhelev) [reply pkg/kubelet/stats/cri_stats_provider_test.go:424]: this definitely makes review harder. Like this line wasn't changed, but still highlighted:

image

**Comment by SergeyKanzhelev** [comment]
> /retest

those seemed to be a valid issues. For package name:

from `verify-pkg-names.sh`

>   echo "!!! Some package aliases break go conventions."
>   echo "To fix these errors, do not use capitalized or underlined characters"
>   echo "in pkg aliases. Refer to https://blog.golang.org/package-names for more info."

As for test failure, as @dashpole mentioned perhaps the timestamp needs to be returned by fake.

Also what I cannot understand, there is already check for the pod's stats timestamp here:

https://github.com/kubernetes/kubernetes/blob/ac97b2d65e6aad6c472b6699b1474a0c0fd276b7/pkg/kubelet/stats/cri_stats_provider_test.go#L746

I wonder if it needs to be updated once fake would start behaving.

**Comment by Zheaoli** [comment]
@SergeyKanzhelev @dashpole Actually I think we don't need mock on the https://github.com/kubernetes/kubernetes/blob/fd74333a971e2048b5fb2b692a9e043483d63fba/pkg/kubelet/stats/stats_provider_test.go#L503

the issue code is here 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L467

I want to change it to `cs.CPU.Time` and I think we have already had the mock data, take a look at this https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider_test.go#L619

But the assertion in my new test will fail on the online test, I think it's weird

File: pkg/kubelet/stats/cri_stats_provider_test.go [hunk 3]
```diff
@@ -28,6 +28,7 @@ import (
 	cadvisorfs "github.com/google/cadvisor/fs"
 	cadvisorapiv2 "github.com/google/cadvisor/info/v2"
 	"github.com/stretchr/testify/assert"
+	assertModule "github.com/stretchr/testify/assert"
```
Comment (dashpole) [inline pkg/kubelet/stats/cri_stats_provider_test.go:31]: looks like we already import this.  Also, `assertModule` is camelcased, which imports shouldn't be.

File: pkg/kubelet/stats/cri_stats_provider_test.go [hunk 4]
```diff
@@ -457,6 +458,8 @@ func TestCRIListPodCPUAndMemoryStats(t *testing.T) {
 	assert.Nil(c1.Logs)
 	assert.Nil(c1.Accelerators)
 	assert.Nil(c1.UserDefinedMetrics)
+	assert.True(assertModule.ObjectsAreEqual(p0.CPU.Time, metav1.NewTime(time.Unix(0, containerStats1.Cpu.Timestamp))) || assertModule.ObjectsAreEqual(p0.CPU.Time, metav1.NewTime(time.Unix(0, containerStats0.Cpu.Timestamp))))
```
Comment (dashpole) [inline pkg/kubelet/stats/cri_stats_provider_test.go:461]: While we are debugging, can you make the error messages more detailed?  Perhaps stick to the form:
`Got %v, want %v`, and print out p0.CPU.Time, cs.CPU.Time and cs.StartTime if it isn't included in the Got, want print statement.

File: pkg/kubelet/stats/cri_stats_provider_test.go (hunk 3 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/stats/cri_stats_provider_test.go:31]: +1 on name. The reason it is renamed is that we declare variable with the same name `assert` and cannot use `assert.ObjectsAreEqual` method after this - variable overrides the package name

**Comment by Zheaoli** [comment]
I will close this PR and reopen it when I solve this problems

**Comment by SergeyKanzhelev** [comment]
I'm sorry with your experience implementing it. It feels like an important issue to fix. Are you actively working on it? If not - please let me know so I can track resolution for it.

**Pull request closed without merging**